backend/tests/data_access/test_repository.py

```diff
@@ -108,6 +110,75 @@ def test_get_artefacts_by_family_name_latest(
     # Assert
     assert len(artefacts) == len(expected_artefacts)
     assert {
-        (artefact.name, artefact.stage.name, artefact.created_at, artefact.version)
+        (
+            artefact.name,
+            artefact.stage.family.name,
+            artefact.stage.name,
+            artefact.created_at,
+            artefact.version,
+        )
         for artefact in artefacts
     } == expected_artefacts
+
+
+def test_get_artefacts_by_family_charm_latest_architectures(
+    db_session: Session, generator: DataGenerator
+):
+    """For latest charms, latest artefacts should include all unique architectures"""
+    # Arrange
+    base = SimpleNamespace(
+        expect=True,
+        name="name-1",
+        version="1",
+        family="charm",
+        track="track-1",
+        stage="edge",
+        day=0,
+        arches=["arch-1"],
+    )
+    specs = [
+        # Base case
+        SimpleNamespace(),
+        # Family is not Charm
+        SimpleNamespace(version="2", family="snap", expect=False),
+        # Name different from base (same version)
+        SimpleNamespace(name="name-2"),
+        # Track different from base (same version)
+        SimpleNamespace(track="track-2"),
+        # Version different from base (same date, return both)
+        SimpleNamespace(version="3"),
+        # Stage different from base
+        SimpleNamespace(version="4", stage="beta"),
+        # Older release than base
+        SimpleNamespace(version="5", day=-1, expect=False),
+        # Older release than base with unique arch
+        SimpleNamespace(version="6", day=-1, arches=["arch-1", "arch-2"]),
+    ]
+
+    expected_artefacts = []
+    for spec in specs:
+        spec = SimpleNamespace(**{**vars(base), **vars(spec)})
+        artefact = generator.gen_artefact(
+            spec.stage,
+            family_name=spec.family,
+            name=spec.name,
+            version=spec.version,
+            track=spec.track,
+            created_at=datetime(2024, 1, 1, tzinfo=timezone.utc)
+            + timedelta(days=spec.day),
+        )
+        for arch in spec.arches:
+            generator.gen_artefact_build(
+                artefact,
+                arch,
+            )
+        if spec.expect:
+            expected_artefacts.append(artefact)
+
+    # Act
+    artefacts = get_artefacts_by_family(db_session, FamilyName.CHARM)
+
+    # Assert
+    assert sorted([a.id for a in artefacts]) == sorted(
+        [a.id for a in expected_artefacts]
+    )
```

Comment:
Thanks for adding comprehensive testing. Would it be possible to simplify this test case by splitting it such that each test is checking one particular case at a time?

Reply:
Now split to better identify what is being tested.